colexec: derive a tracing span in materializer when collecting stats

yuzefovich opened the pull request:
If we don't do this, then the stats would be attached to the span of
the materializer's user, and if that user itself has a lot of
payloads to attach (e.g. a joinReader attaching the KV keys it looked
up), then the stats might be dropped based on the maximum size of
structured payload per tracing span of 10KiB (see
`tracing.maxStructuredBytesPerSpan`). Deriving a separate span
guarantees that the stats won't be dropped.

This required some changes to make a test - that makes too many
assumptions about tracing infra - work.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/81995)
<!-- Reviewable:end -->

yuzefovich commented:
Here is an example of TPCH Q9 [before](https://cockroachdb.github.io/distsqlplan/decode.html#eJzEV9tu20YTvv-fYrFAYMn_2uLyIFIEAtBxncSND4HspAgKg6DFicKap-wuUQtBHqsv0CcrlgeJokjK1EWrG5EznJ3DNzP8-APz7yG28d3F1cX5PYo9ESQxQYm7Ao8RxLNo5EVJFosxOruTt27Kkq-BQG_nt9doVJm5sReBfKI6AJ4F8xZidCTPOZIHJswH5nsC8pMqB6PQhWcBsQ9-yoIFoGM0ougEha4f8EXueIxO0CjlLs_SNFwtEi7QMQrd75kXi0Cs8uOKGIugUo8JGXmahgEwgsIghkBARHKNlBOUB8NJGS767f3F_AKN8l_h6AlW6DUKq-sxOrv5ZR1FU7nRSg-Vtrxea_uUZXkqbXVTqblbBFro483d5vAcgKvLDxfo6NWSAcSvjsZ5aUq83s1vP31Eb740IMYEx4kPN14EHNu_Y4ofCE5ZsgDOEyZFP_IHLv1nbCsEB3GaCSl-IHiRMMD2DywCEQK28b33GMIcPB_YRMEE-yC8IMyPrdBwuBs_YYLPkzCLYm6jdUEJqiWJCb5LPak_wQR_-IxEEIGNqK5FvBAsklhAnINX6JS__ypVLPmTIwaebyOqEEVRCvHjSkAlV2ca-hC8wQRffz4_R1xAivJmQyN4FpMgFuPS2JgUf-tHAZ5aHlUnlGKCI-8ZRRAlbIW8MEwWngDfRpqmlN4ePbH4BhwlmUgzIV1ggvN4N5I84oefBBeistZceEvANv1JDsODbuNRTYKTsiDy2KoOyaaNCUpr-GyNYAdCpmEMQ8hSOiCiGkXX-yAqrI1J-d8PkmZNNLMTJKqU7poYmZbaBKkKugsl9VCU1F2UWhGqAZQPfgcaqjFwXtQuNFRrPxhqCYb6EjDodEKN4WDQ2bQJhroHDG0IGG-DUAADNtG2kSjkNnLUxo61bfvufn558w4TDM-wyGolLor_ghSoQqZTvTMDvTODTeBZnL8zwN-K-0Fa7nukpQzXwJbwaxLEwCb6diVC-CpGDv3_-DULlt-KS0zwrczDocRRiaMRRyeO0VIRK69Ix5bUqy0p9fx7iAREKfID_oQy7i1lz6K2njBos566SqbGtLOexpCOkFUop9NoVKLkFU7opk8ul6-1qyR5ylL0RxLEKInLgtQnd_NqJxsSQDZUQl7Kc-MsegTWmcH0wAymHRlsdsxOBlPizLYzqIiXjHWLu5Eaa1u3hOwEIk8xiWMRhyrEoZQ4VCVVt3RmaR6YpbmdZcH1-nLU6gkmNYhqrBWX7vN1aSONzoiu6GspF14YVjRldqrzrT2snFqD9rA8XVHNlrfibLafuCg5--pZvsUDzeFU6anWPZ7qqdGxkzV1ZwCr6nQha_3rG-29x7-VC81qXWjrfWZ0rLNaE88afawRh-rEoW0rj5pGz9Kb6mVRB-48jVpDaz47lJTMtutVkPQ2WlL7LiHlJ2EXkTenA4mJ0cJJTq39s6ApE23PNPSTd7WTu-8QEaObtiv_ZcNTpa3j1c0bXHb8XO5xZjc-3Nf0Rnb3mKCRMzse0VPFtu23V7dn99aJQ5Xx-GTkUO3YscYEOdRsmQJT6RkC1TrwzX_AFNBBX1BnyyWDpScSVrRIrYhyLUjR2c0X9-b23r35dHU1crTxjiiv7t2na1nx1q3bx4nMF3cfNXvab9AHyRx4msQcmm3YerQiew_8JRS9zJOMLeAjSxa5m-L2NrfLBT5wUWit4uYyzlU5KnVj2mus9xurvcbalrHSNNYGeFabxnqvsdHv2eg1nvYbT3uNzX5jcwBUOzlbvcZU6cdqNsR6xzftbzJK-_Om_W1G1Yb5w8___RMAAP__hFLNPA==) and [after](https://cockroachdb.github.io/distsqlplan/decode.html#eJzEWP1u2zYQ_39PQRAoYmeMLVLfAgoozdI2az4KJ-1QDIGhWKyrRV8lKSxG0cfaC-zJBurDlmVJroxsyz-R7nTk8X6_4935G-RfQ-jA2_PL87M7EHsiSGIEkvmKegwBnkUjL0qyWIzB6a18nacs-RwI8Hp2cwVGldk89iIqv6gWoE-CeQsxOpLrHMkFE-ZT5nuC5itVG4zCOX0SNPapn7JgQcExGGFwAsK5H_BFvvEYnIBRyuc8S9NwtUi4AMcgnH_NvFgEYpUvV_hYOJV6TEjP0zQMKEMgDGIaCBqhXCPlCOTOcFS6C357ez47B6P8r9joka7ASxBWz2Nwev3L2oumcqOVO1Ta8nmt7VOW4am01Uul5vPC0UIfb942i-cAXF68OwdHL5aM0vjF0TgPTYnXm9nNh_fg1acGxBDBOPHptRdRDp3fIYb3CKYsWVDOEyZF3_IPLvwn6CgIBnGaCSm-R3CRMAqdb1AEIqTQgXfeQ0hn1PMpmyoQQZ8KLwjzZSs0XD6PHyGCZ0mYRTF3wDqgCNQOCRG8TT2pP4EIvvsIRBBRB2BNjXghWCSxoHEOXqFT_v6rVLHkTw4Y9XwHYAUpilKIH1aCVnJiq-Bd8AoiePXx7AxwQVOQkw2M6JOYBrEYl8b6tPi3_pTSx5ZPyRRjiGDkPYGIRglbAS8Mk4UnqO8AVVXK3R48sfhCOUgykWZCbgERzP3dSHKP778jWIjKWHPhLSl08Hd0GB54G48qE9yUBZHHVnVINjRGIK3hs5WCHQgZOhmGkKV0QIRVDK72QVRY69Pyfz9IqjVVzU6QsFJu18TItEgTpMrpLpTIoSiRXZRaEaoBlCd-BxrE1oahQbrQINZ-MEgJBvkRMLAxxfpwMLBtNMEge8BQh4DxOggFZZRN1W0kCrkDXNK4Yx3Hub2bXVy_gQjSJ7rIaiEuLqsfOAJWkGFonSfQOk-wcTyL85pB_S2_76Xlvk9awnBF2ZL-mgQxZVNtOxIh_SxGLv55_JIFyy_FI0TwRp7DxcglyFWRqyFXb4mIlUek45bUqltS6vnXEAgapcAP-CPIuLeUnAVtnNBxM54aQYZudMZTH8IIGYUyO_VGJMq-wg3n6eOcy7J2mSSPWQr-SIIYJHEZkHrmbko72jQBaNNKyEe5bpxFD1QW59y1PEXXx6qEXHhhWAbWNvQq0wuBPijvVWwjTdFabuGJuT_xlbza9yR78UGTDPoEd7OBTOyOK0AlO3hX7ncBbhwIuNEB-OZK3gHcQK69DXjVp0pot1pdVGty1xkkEwfJVUzkWsjFCnIxRi4mqEquBik20O2ywigivOmfyIQ8Dy-IauxvoA7kBSYTq48YxvMRwzyQGOY2MYppoo8Wap0TSe0SqM1FA4DF1sTeQpYoRcwGIasQsyXjbftfQ5bggn9dyOrPh6z1n9fMtx7_UpZMq7Vkrium3lEwa3lvN1JfRS7WkIvbiio27Z6yamhlUAdWVRVbQ2NuH9r22tvxKsbAtsa3Nvmi8keHrlFRGTiIEL3llptY-3NBVabqnmzoHw9J53S40-rq3YOh8n8SHittjCebHlEyfiZLH3MaPw2tG2jJ7jECI9c-HuGJ4jjO68ub0zvrxMXKeHwycrF67FpjBFxstmSBqfQkAbEO7C0PyAI8aEY_XS4ZXXoiYQVFakGU14IUnV5_ml_f3M2vP1xejlx1vCPKo3v74UpGvPXW7eu6zR9mHzZ76Ddo5J1RniYxp00ati6tSO5Rf0kLLvMkYwv6niWLfJvi9Sa3ywU-5aLQWsXLRZyrclTqxrjXWOs3Jr3G6pax0jRWB-xMmsZar7Hev7Pea2z0Gxu9xma_sTkAqp0zW73GWOnHyh5ivbM37icZxv3nxv00w6Rhfv_9p38CAAD__yBCh8k=) the fix. Note how `JoinReader/5` and `JoinReader/6` are missing stats. Confusingly, the behavior is non-deterministic, at least partially (probably because the traced events in KV or in storage are different).

msirek reviewed (approved):
Nice fix!
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81995#-:-N3RQ-kl5ZMC_PESi4YL:b8ig6vo)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cucaroach and @yuzefovich)

___
*[`pkg/sql/colexec/materializer.go` line 257 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/81995#-N3RNFmfEH0FKfU4rXUk:-N3RNFmfEH0FKfU4rXUl:b-xr9t5w) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c953ca331480b8db3542f8b4ffcf17c9ac19009/pkg/sql/colexec/materializer.go#L257)):*
> ```Go
> 	} else {
> 		ctx = m.StartInternalNoSpan(ctx)
> 	}
> ```

The code changes look good, but I have a question. The `RowSource` interface says:  "RowSources that consume other RowSources are expected to Start() their inputs. "  Is there are reason that `Materializer.Start` doesn't call `Start` on its `input`?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
The reproduction here requires a lot of data, so I think it's ok to merge without a regression test. There isn't really any downside to this change too (having an extra tracing span incurs some overhead, but when we're collecting stats, we don't really care about that).

TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81995#-:-N3RWkkhBbOVTFaaqCio:be6jgr1)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cucaroach and @msirek)

___
*[`pkg/sql/colexec/materializer.go` line 257 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/81995#-N3RNFmfEH0FKfU4rXUk:-N3RWB6X7IHwNW37XlTS:b-xhsj98) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c953ca331480b8db3542f8b4ffcf17c9ac19009/pkg/sql/colexec/materializer.go#L257)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

The code changes look good, but I have a question. The `RowSource` interface says:  "RowSources that consume other RowSources are expected to Start() their inputs. "  Is there are reason that `Materializer.Start` doesn't call `Start` on its `input`?
</blockquote></details>

`Materializer` is an adapter from the `colexecop.Operator` interface to the `RowSource` interface. `Operator.Init` is roughly equivalent to `RowSource.Start`, and a few lines below, wrapped in a panic catcher, we do call `Init` on the input. The adapter in the opposite direction (from the `RowSource` to `Operator`) is the `Columnarizer`, and `Columnarizer.Init` calls `input.Start`.


<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81995)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cucaroach and @yuzefovich)

___
*[`pkg/sql/colexec/materializer.go` line 257 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/81995#-N3RNFmfEH0FKfU4rXUk:-N3RXJJEAP5svVYnfdfB:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c953ca331480b8db3542f8b4ffcf17c9ac19009/pkg/sql/colexec/materializer.go#L257)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

`Materializer` is an adapter from the `colexecop.Operator` interface to the `RowSource` interface. `Operator.Init` is roughly equivalent to `RowSource.Start`, and a few lines below, wrapped in a panic catcher, we do call `Init` on the input. The adapter in the opposite direction (from the `RowSource` to `Operator`) is the `Columnarizer`, and `Columnarizer.Init` calls `input.Start`.
</blockquote></details>

:thumbsup: 


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Looks like the newly added tracing span is being used after it was finished, I'll need to look into that.

bors r-

yuzefovich committed:
colexec: derive a tracing span in materializer when collecting stats

If we don't do this, then the stats would be attached to the span of
the materializer's user, and if that user itself has a lot of
payloads to attach (e.g. a joinReader attaching the KV keys it looked
up), then the stats might be dropped based on the maximum size of
structured payload per tracing span of 10KiB (see
`tracing.maxStructuredBytesPerSpan`). Deriving a separate span
guarantees that the stats won't be dropped.

This required some changes to make a test - that makes too many
assumptions about tracing infra - work.

Release note: None

yuzefovich reviewed (commented):
I had to make some tweaks to how the materializer is closed. Please take another look.

The problem was that some of the closers would try to log an event into the already finished tracing span (which happens automatically in `InternalClose`, which was previously done before closing the closers), so this would lead to use-after-finish crash with tracing infra. Some examples of such closers are `rowexec.orderedAggregator.close`, `rowexec.hashAggregator.close`, `colexec.externalSorter.Close`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81995#-:-N3XOQ584CmBF1h0qu9U:b-jcbxj2)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @msirek)

___
*[`pkg/sql/colexec/materializer.go` line 257 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/81995#-N3RNFmfEH0FKfU4rXUk:-N3XOG0Q1tSTDrNLxNVh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0ac51d7333b1b5f38f71ea4d93b5c13eb6e0238/pkg/sql/colexec/materializer.go#L257)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

:thumbsup: 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
Thanks for the explanation. The updated code looks good.

> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81995#-:-N3XyMV5-0D788Nl6LuQ:b4zeng9)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @msirek)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
This change doesn't seem worth backporting given that it's not exactly trivial, so I think we'll just merge it on master.

TFTRs!

bors r+

yuzefovich commented:
Hm, looks like this change slows down the tests somehow so that they time out (or more likely to time out). I'll need to take a closer look at that.

yuzefovich commented:
I think I figured it out - `TestAlreadyRunningJobsAreHandledProperly` relies on how the traces work, and this change was breaking those assumptions. I added some testing knobs to make those assumptions true again.

Since I only modified the testing knobs behavior since I got LGTMs, I don't think extra reviews are needed.

The only failure I saw on CI is #85934 which is present on master.

bors r+

yuzefovich commented:
bors r+
